### ISSUE TITLE
UI design for Let's Encrypt cert renewal and use by other modules

### ISSUE THREAD

**Issue by JoKeyser** [freedombox/Plinth#1091]
This is to discuss how to implement this issue @JosephKiranBabu raised in PR #1073:

> [...] we should rethink the strategy of automatic certificate renewals. In particular, not asking user input for certificate renewals and removing the special status given to ejabberd on the letsencrypt page, since diaspora and other applications which would need subdomains will compete for that space on that page.

**Comment by JoKeyser** [comment]
Here's a screenshot the current page:
![lecurrent](https://user-images.githubusercontent.com/8722846/31862563-6ed7a896-b72f-11e7-9b69-0cc75d7d6a4c.png)

**Comment by JoKeyser** [comment]
@JosephKiranBabu touches upon 3 sub-issues, here my opinions:

1. Not asking user input for certificate renewals: (The option adds/removes Plinth as hook script that gets called when certbot renews a certificate.) I agree that renewal through Plinth should be the default. But should there be an option to disable it? I guess the use only (weak) case would be to give an easy option to (advanced) admins. But such an advanced admin might just as well edit certbot's config file directly. So I think removing the option entirely is justified. 
2. Removing the special status of ejabberd: I agree, similar reasoning as above: Default and get rid of the option. Any module that is able to use an LE cert would then just do so, as soon as it's obtained. I see no real benefit of optionally sticking with the self-signed certs. The only sensible choice in the future might be between Monkeysphere and LE?
3. Diaspora and other apps that need subdomains: I've got no clear opinion here; Are subdomains really necessary? To me, they seem to unnecessarily complicate things (for our use-case).

**Comment by JoKeyser** [comment]
Another thing I'm unclear about is whether LE cert renewal and use other modules should pertain only ot the current domain, as currently done? Is there a good reason to keep cert management for multiple domains?
My (strong) opinion here is to keep it simple for everyone, by sticking to just one - or am I overlooking something?

**Comment by mray** [comment]
Wow. This screen looks like coming fresh from admin hell.

> My (strong) opinion here is to keep it simple for everyone...

I strongly agree. Can you put it in very simple words what people are supposed to do on that page in the first place? My current understanding is that they can make choices about encryption capabilities for their services. in that case sensible defaults should go a long way, if more fine-grained decisions are needed they need to be more approachable.

**Comment by jvalleroy** [comment]
@JoKeyser It is possible for a FreedomBox to have more than one name, for example and Dynamic DNS domain name and a Pagekite name. But I think the default behavior could be to just acquire (and renew) Let's Encrypt certificates for all supported names.

**Comment by JoKeyser** [comment]
@mray thanks for the clear words :). I think the page is the result of adding its features non-disruptively, hence all is optional. Now we should define here what the sensible defaults are and clean it up. The LE page's purpose is to let people *manually* get (and revoke, and delete) LE certificates - instead of using self-signed ones that elicit warnings in clients.

Also, what I forgot so far is: Since PR #988, LE certs are attempted to be obtained automatically once the domain changes (and the old one revoked). That jumps ahead of the current LE page's request to read and agree to the [(7-page-long) Let's Encrypt Subscriber Agreement](https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf). Maybe we should put the link to this then also on the domain-configure page?

**Comment by JoKeyser** [comment]
@jvalleroy thanks for clarifying; DDNS and Pagekite are somewhat out of my radar, I should try them too. So what best to do with the different services, e.g. ejabberd, mumble, etc? Should they also be available on all these names (do they all tolerate this?), or should there be a "main" domain?

**Comment by mray** [comment]
This is what comes to my mind from a UI perspective:


## 1.
> ![image](https://user-images.githubusercontent.com/164785/31865784-a3e76f22-b774-11e7-8f18-108ed3c9052c.png)

This should be replaced by one or two simple sentences like:  
"Lets Encrypt allows you to enable, disable or revoke SSL encryption capabilities of your services"
When removing important info, we should make sure it is present in the respective WIKI section.


## 2. 
> ![image](https://user-images.githubusercontent.com/164785/31865821-3f2a5ac6-b775-11e7-9d6d-f2ade8fb3bb6.png)
 
This list could look more like this:

![image](https://user-images.githubusercontent.com/164785/31866072-a9203ff0-b779-11e7-9d8c-2fd56c5df3f6.png)


##  3.
The "Run Diagnostics" button should be moved out of the way, I propose to the very bottom.

## 4.

>  ![image](https://user-images.githubusercontent.com/164785/31866090-02711188-b77a-11e7-8525-6c411c4ebde2.png)

I'm not sure for what reason this would be relevant any more. If the last option relates to ejabberd only, it does not belong to this module, but to ejabberd.

**Comment by JosephKiranBabu** [comment]
@JoKeyser 

1. *Not asking user input for certificate renewals:* We are going to introduce Cockpit into FreedomBox for advanced system administration use cases. Whenever we think of adding such features we should first check whether Cockpit has it already. The integration is close enough now that we can probably release it in the next version of Plinth. Our Twitter account advertises FreedomBox as "setup once an forget about it". Our certificate renewals should be a one-time setup in my opinion.
2. *Removing the special status of ejabberd:*  I agree with what you wrote.
3. *Diaspora and other apps that need subdomains:* Subdomains are unfortunately necessary for Diaspora and some other applications which don't like to run behind a reverse proxy. 

Regarding the distinction between current-domain and non-current domain, I agree with what @jvalleroy said. We need to be able to support certificates for all kinds of domains (except .onion). I don't have a strong opinion on revoking old certificates. In theory, multiple domain names can point to the same IP address, but we aren't currently supporting that use case, since our Configure page only gives one option for domain name.

@JoKeyser I have to point out one issue with Diaspora and Matrix. They do not tolerate the domain name being changed after initial setup. So, automatically revoking the cert for their domain names (because of some condition like it's not the *current* domain name) might cause unnecessary problems.

One more thing we have to keep in mind is, whether spending a lot of effort on maintaining multiple certificates for each FreedomBox is worthwhile since Let's Encrypt is going to support wildcard certificates from January.

@mray I think the whole "Certificate Management" section shouldn't be the user's responsibility at all. FreedomBox should manage it's own certificates as much as possible. We can still provide some user interface for manual overrides and diagnostic purposes.

**Comment by mray** [comment]
@JosephKiranBabu do you think the "Certificate Management" section will be obsolete soon?

**Comment by JosephKiranBabu** [comment]
@mray Yes. I think it's unnecessary maintenance work for the user. When we expect people to use FreedomBox like a home appliance, we cannot expect them to know what certificates are.

There are two options in that section.
1. _Let FreedomBox manage the certificate renewal for this domain._ 
FreedomBox will do this by default, so no user decision required.
2. _Use certificate of the current domain for ejabberd._ 
I'm not fully in context of what the problem with ejabberd is, but I think we can use the same domain name configuration screen that we're using with Matrix, Diaspora and Tahoe-LAFS. So, this option moves to the Setup page of ejabberd. The user simply selects a domain name and FreedomBox takes care of obtaining and renewing certificates for it.

**Comment by JoKeyser** [comment]
Okay, I'm getting a confused by plethora of things to consider here :). I'll ask some questions to clarify this (at least for me). I think I'm confused by (sub-)domains vs certificates, and the difference what *currently is* the case and how it *should* be.

* Q1: What are all the current (and concretely planned) mechanisms/forms by which (sub-)domains can be added?
* Q2: Should FreedomBox (try to) obtain a Let's Encrypt certificate be obtained in *each* case of adding a domain?
    * Q2a: Shouldn't obtaining a LE cert be always be *optional* (but the default), rather than mandatory?
* Q3: Should all modules (ejabberd, mumble, synapse,...) offer a choice of domain (and certificate) for their operation?
* Q4: Should the *use by a module* of an obtained certificate for a given domain be optional (default yes), or mandatory?
    * Q4a: Should a module in general be able to create a (sub-)domain for itself and obtain the certificate, or should it only be able to use existing ones? (Diaspora being one exception that requires it currently, but should this be a general design?)

**Comment by JoKeyser** [comment]
> @JosephKiranBabu 
>   1. *Let FreedomBox manage the certificate renewal for this domain.*
>    FreedomBox will do this by default, so no user decision required.

I agree with setting the default to simplify user decisions, but I think there should be one *offered*. As I mentioned earlier, it may even be necessary to have the option, because of the formal Let's Encrypt Subscriber Agreement. I feel it would be too intrusive to "either use Let's Encrypt or don't use FreedomBox" :).

**Comment by JoKeyser** [comment]
@JosephKiranBabu I agree with you on (not) revoking old certificates. Revocation is a relatively drastic operation that should be only manually triggered.

**Comment by JoKeyser** [comment]
@JosephKiranBabu , @mray 
> 2. *Use certificate of the current domain for ejabberd.*
> I'm not fully in context of what the problem with ejabberd is, but I think we can use the same domain > name configuration screen that we're using with Matrix, Diaspora and Tahoe-LAFS. So, this option moves to the Setup page of ejabberd. The user simply selects a domain name and FreedomBox takes care of obtaining and renewing certificates for it.

Just to clarify: The special status of ejabberd was always intended to vanish. I put this entry on the Let's Encrypt page because it would be about more modules that use the certificate in the future, not just ejabberd.
Now the question is whether the choice to use an (existing) LE-cert is needed at all, and if so, is it needed on a per-module-basis. (Q4, above). And then how an UI should look like - is it on each module's page, or an overview on the LE page, or both?

**Comment by JoKeyser** [comment]
@JosephKiranBabu thanks for pointing out that wildcard certs are [coming in January 2018](https://letsencrypt.org/2017/07/06/wildcard-certificates-coming-jan-2018.html). With only 6 weeks to go, I think it makes sense to write all new code already as if it was there. It surely will make things easier.

**Comment by SunilMohanAdapa** [comment]
> Q0: Should FreedomBox users think more in terms of one domain for all services
> (with a few unfortunate exceptions like Diaspora, which we try to fight
> against), or several domains (selectable per module/service)? The first
> emphasizes simplicity, the second freedom of choice. (I personally lean
> towards the first, at least short/midterm. That's why the current
> implementation emphasizes the currently set domain so much. Both can be
> combined by setting defaults toward the first, and ability to handle more
> complex setups (at the cost of making the implementation more complicated).)

I have written about the need for subdomains below for another questions.

Freedom of choice is not a technical or political goal for our project. If we
make a decision to support multiple domains it should be based on other reasons
and not for providing the user with freedom of choice. Currently the main
reasons for providing multiple domains are anonymity (.onion addreses),
reachability (Dynamic DNS, Pagekite), technical limitations (subdomains). More
below.

I believe also that handling multiple domains does not complicate the
implementation too much. If we consider the cases of user changing the domain
name, we already have to handle removing, renaming and adding domains. This is
where most of the complexity is.

> Q1: What are all the current (and concretely planned) mechanisms/forms to add
> (sub-)domains?

- Pagekite (always a sub domain of pagekite.net or some other service)
- Statically adding domains (in Configure page, possibly in DNS server later)
- Tor (.onion addreses)
- Dynamic DNS (as subdomain of an existing domain such as freedombox.rocks).
- Subdomains (such as in case of Diaspora)

> Q1a: What are current/planned ways domains can be deleted? One current
> confusion I have is when changing the domain name on the Configure page, then
> previously obtained LE certificates remain (and will remain to get renewed).

In this case, when domain is removed from the system, its let's encrypt
certificate should be removed as well since we will keep on attempting to renew
it otherwise.

> Q2: Should FreedomBox (try to) obtain a Let's Encrypt certificate be obtained
> in each case of adding a domain?
>
> Q2a: Shouldn't obtaining a LE cert be always be optional (but the default),
> rather than mandatory?

I believe there is no use case for offering insecure services when they can be
secured using lets encrypt. Even when offering no services on a particular
domain, we can obtain a certificate for it. It keeps things simple. If someone
truly owns a domain, they also have the right to obtain certificate for it (LE's
view).

Offering such options complicates the UI for the users and code for developers
without much benefit. The maximum we should do here is to provide an option to
completely disable LE.

If LE is implemented properly, everything automatically works and user will
never have to come to LE page or worry about what certs are.

> Q3: Should all modules (ejabberd, mumble, synapse,...) offer a choice of
> domain (and certificate) for their operation? Currently, some offer the choice
> (from existing domains), and some just take it from the currently set domain.

There should not be a 'currently set domain'. By this we mean a default domain
which is set in the configure page. Domain set here should not have priority
over other domains such as Dynamic DNS domains.

Many applications can actually work on all or multiple domains (such as
transmission, tt-rss). They should not ask for which domain to work on.

Some applications work on a single domain but can change their domain at any
time. Here one of the domains can be assumed and configured (this can be the
domain added from Configure page). Later an option may be provided to switch the
domains.

Some applications work on a single domain and can't be changed later. Here we
must allow the user to configure a domain before setting up the application.

> Q3a: Should modules in general be able to create a (sub-)domain for itself and
> obtain the certificate, or should it only be able to select from existing ones
> (see Q1)? (Diaspora being one exception that currently requires creation, but
> should this be a general design?)

Assuming we have a robust way of creating and handling subdomains, there are
several advantages to using subdomains instead of URL fragments.

- Protocols such as Diaspora's protocol assume that some API for federation is
  fixed. They don't use discovery mechanisms such as .well-know to first
  discover the API URL and then use it. All such applications will continue to
  work.

- Cookies are not shared across subdomains but are shared across URL fragments
  by default. This means that applications hosted on multiple subdomains are
  more secure.

- Many applications such as Cockpit are easier to configure on a subdomain than
  they are on a URL fragment. We will be simplifying development effort for
  writing and maintaining applications for FreedomBox with a proper subdomain
  framework.

- For many applications, installing on URL fragment is a less tested code path
  leading to bugs which may be specific for FreedomBox users.

While we are not currently looking to host each application as a subdomain, we
should not reject this possibility. In case, an application should be allowed to
create a subdomain and request certificate for it. The part of requesting
certificates will get simplified when LE has support for wildcard certificates.

BTW, obtaining wildcard certificates seems to be initially limited to DNS based
verification. Unless FreedomBox has full DNS control over the said domain, it
will not be possible for us to obtain such certificates easily.

> Q4: Should the use by a module of an obtained LE certificate for a given
> domain be optional (with default yes), or mandatory (if it exists)? In other
> words, does it make sense to offer the choice to not use an existing LE
> certificate? Currently, ejabberd is the only module, offering the choice to
> use it or not (default)

There is no use case to allowing an application to work insecurely. On a
per-application basis, we can decide if such a need arises (or if there is an
error in obtaining or using the certificate) and provide the control in the
application.

Summary
==

- Subdomains are needed.
- Multiple domains are needed.
- Remove certificates for deleted domains.
- Always obtain cert for a new domain automatically.
- Don't provide option to turn of automatic renewal.
- Don't provide options to disable certificates on a per-application basis.
- Maybe provide an option to turn of LE entirely.

**Comment by JoKeyser** [comment]
@SunilMohanAdapa huge thanks to clarifying and answering all these; I'm fine with all these choices, and finally understand why subdomains are desirable.

I think the option to turn LE off entirely is necessary, and that's the place where the LE license agreement should also be placed. Maybe nitpicking, but in my eyes the choice is not between using-LE vs "insecure", but between using-LE vs. self-signed cert - of course that can be considered practically insecure because nobody really checks the certs, so enabling LE should clearly be enabled by default. Proposal (placed at the Configure page?):  
* [x] Let FreedomBox obtain and use Let's Encrypt certificates for my domain(s), and I agree with the [Let's Encrypt Subscriber Agreement](https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf). *Enabling this option allows clients and other servers to automatically authenticate your FreedomBox cryptographically (recommended).*
